### PR TITLE
Feature/separate instance rule

### PR DESCRIPTION
**Pull request by matsuoka-k-git**
# Description
* Add nifcloud_separate_instance_rule resource.

## How Has This Been Tested?

- [x]  Buld with `make install` command, and Apply example.

```:console
make install
cd examples/separate_instance_rule
terraform init -plugin-dir ~/.terraform.d/plugins
terraform plan
terraform apply
```

- [x] Run acceptance tests.

```:console
TF_ACC=1 go test ./nifcloud/acc/... -v -count 1 -parallel 20 -timeout 360m -run TestAcc_SeparateInstanceRule
```

## Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (improves code quality but has no user-facing effect)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

### CODE REVIEW

File: nifcloud/resources/computing/separateinstancerule/update.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+		req := svc.NiftyUpdateSeparateInstanceRuleRequest(input)
+
+		_, err := req.Send(ctx)
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("failed updating separate instance rule description: %s", err))
+		}
+	}
+
+	if d.HasChange("name") && !d.IsNewResource() {
```
Comment (tunakyonn) [inline nifcloud/resources/computing/separateinstancerule/update.go:26]:
```suggestion
	if d.HasChange("name") {
```

File: nifcloud/resources/computing/separateinstancerule/delete.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/nifcloud/nifcloud-sdk-go/nifcloud"
+	"github.com/nifcloud/nifcloud-sdk-go/service/computing"
+	"github.com/nifcloud/terraform-provider-nifcloud/nifcloud/client"
+)
+
+func delete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	svc := meta.(*client.Client).Computing
+
+	req := svc.NiftyDeleteSeparateInstanceRuleRequest(&computing.NiftyDeleteSeparateInstanceRuleInput{
```
Comment (tunakyonn) [inline nifcloud/resources/computing/separateinstancerule/delete.go:17]: Use expandNiftyDeleteSeparateInstanceRuleInput.

File: nifcloud/resources/computing/separateinstancerule/update.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+		_, err := req.Send(ctx)
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("failed updating separate instance rule name %s", err))
+		}
+
+		d.SetId(d.Get("name").(string))
+
+	}
+
```
Comment (tunakyonn) [inline nifcloud/resources/computing/separateinstancerule/update.go:39]: Please implement change instance_id and instance_unique_id.

File: nifcloud/resources/computing/separateinstancerule/read.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	input := expandNiftyDescribeSeparateInstanceRulesInput(d)
+	svc := meta.(*client.Client).Computing
+
+	req := svc.NiftyDescribeSeparateInstanceRulesRequest(input)
+
+	res, err := req.Send(ctx)
+
+	if err != nil {
+		var awsErr awserr.Error
+		if errors.As(err, &awsErr) && awsErr.Code() == "Client.InvalidParameterNotFound.SeparateInstanceRules" {
```
Comment (tunakyonn) [inline nifcloud/resources/computing/separateinstancerule/read.go:24]: Maybe error responce is mistake.

File: examples/separate_instance_rule/main.tf [hunk 5]
```diff
@@ -0,0 +1,64 @@
+terraform {
+  required_providers {
+    nifcloud = {
+      source = "nifcloud/nifcloud"
+    }
+  }
+}
+
```
Comment (miyuush) [inline examples/separate_instance_rule/main.tf:8]: I want to match it with other examples, so adding a region seems like a good idea.

```suggestion

provider "nifcloud" {
  region = "jp-east-1"
}

```

File: nifcloud/resources/computing/separateinstancerule/expander_test.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+	tests := []struct {
+		name string
+		args *schema.ResourceData
+		want *computing.NiftyUpdateSeparateInstanceRuleInput
+	}{
+		{
+			name: "expands the resource data",
+			args: dn,
+			want: &computing.NiftyUpdateSeparateInstanceRuleInput{
+				SeparateInstanceRuleName:       nifcloud.String("test_separate_name"),
```
Comment (miyuush) [inline nifcloud/resources/computing/separateinstancerule/expander_test.go:101]: Isn't it a "test_separate_name2"?

File: nifcloud/resources/computing/separateinstancerule/expander.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func expandNiftyUpdateSeparateInstanceRuleInputForDescription(d *schema.ResourceData) *computing.NiftyUpdateSeparateInstanceRuleInput {
+	return &computing.NiftyUpdateSeparateInstanceRuleInput{
+		SeparateInstanceRuleName:              nifcloud.String(d.Id()),
+		SeparateInstanceRuleDescriptionUpdate: nifcloud.String(d.Get("description").(string)),
+	}
+}
+
+func expandNiftyRegisterInstancesWithSeparateInstanceRuleInstanceIDInput(
```
Comment (miyuush) [inline nifcloud/resources/computing/separateinstancerule/expander.go:61]: Is there any need to test this method?

File: nifcloud/resources/computing/separateinstancerule/expander.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	d *schema.ResourceData,
+	instanceIDs []string,
+) *computing.NiftyRegisterInstancesWithSeparateInstanceRuleInput {
+	return &computing.NiftyRegisterInstancesWithSeparateInstanceRuleInput{
+		SeparateInstanceRuleName: nifcloud.String(d.Get("name").(string)),
+		InstanceId:               instanceIDs,
+	}
+}
+
+func expandNiftyDeregisterInstancesFromSeparateInstanceRuleInstanceIDInput(
```
Comment (miyuush) [inline nifcloud/resources/computing/separateinstancerule/expander.go:71]: Is there any need to test this method?

File: nifcloud/resources/computing/separateinstancerule/expander.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	d *schema.ResourceData,
+	instanceIDs []string,
+) *computing.NiftyDeregisterInstancesFromSeparateInstanceRuleInput {
+	return &computing.NiftyDeregisterInstancesFromSeparateInstanceRuleInput{
+		SeparateInstanceRuleName: nifcloud.String(d.Get("name").(string)),
+		InstanceId:               instanceIDs,
+	}
+}
+
+func expandNiftyRegisterInstancesWithSeparateInstanceRuleInstanceUniqueIDInput(
```
Comment (miyuush) [inline nifcloud/resources/computing/separateinstancerule/expander.go:81]: Is there any need to test this method?

File: nifcloud/resources/computing/separateinstancerule/expander.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	d *schema.ResourceData,
+	instanceUniqueIDs []string,
+) *computing.NiftyRegisterInstancesWithSeparateInstanceRuleInput {
+	return &computing.NiftyRegisterInstancesWithSeparateInstanceRuleInput{
+		SeparateInstanceRuleName: nifcloud.String(d.Get("name").(string)),
+		InstanceUniqueId:         instanceUniqueIDs,
+	}
+}
+
+func expandNiftyDeregisterInstancesFromSeparateInstanceRuleInstanceUniqueIDInput(
```
Comment (miyuush) [inline nifcloud/resources/computing/separateinstancerule/expander.go:91]: Is there any need to test this method?

File: examples/separate_instance_rule/main.tf (hunk 5 above)
Comment (tunakyonn) [reply examples/separate_instance_rule/main.tf:8]: Fixed!

File: nifcloud/resources/computing/separateinstancerule/flattener_test.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFlatten(t *testing.T) {
+	rd := schema.TestResourceDataRaw(t, newSchema(), map[string]interface{}{
+		"name":               "test_separate_name",
+		"availability_zone":  "test_availability",
+		"description":        "test_description",
+		"instance_id":        []interface{}{"test_instance_id1", "test_instance_id2"},
+		"instance_unique_id": []interface{}{"test_instance_unique_id1", "ttest_instance_unique_id2"},
```
[the remark this answers is not in the document]
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/flattener_test.go]: Fixed!

File: nifcloud/resources/computing/separateinstancerule/flattener_test.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/nifcloud/nifcloud-sdk-go/nifcloud"
+	"github.com/nifcloud/nifcloud-sdk-go/service/computing"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFlatten(t *testing.T) {
+	rd := schema.TestResourceDataRaw(t, newSchema(), map[string]interface{}{
+		"name":               "test_separate_name",
+		"availability_zone":  "test_availability",
```
[the remark this answers is not in the document]
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/flattener_test.go]: Fixed!

File: nifcloud/resources/computing/separateinstancerule/expander_test.go (hunk 6 above)
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/expander_test.go:101]: Fixed!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 7 above)
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/expander.go:60]: Added!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 8 above)
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/expander.go:70]: Added!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 9 above)
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/expander.go:80]: Added!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 10 above)
Comment (tunakyonn) [reply nifcloud/resources/computing/separateinstancerule/expander.go:90]: Added!

File: nifcloud/resources/computing/separateinstancerule/flattener_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/flattener_test.go]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/flattener_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/flattener_test.go]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/expander_test.go (hunk 6 above)
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/expander_test.go:101]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 7 above)
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/expander.go:60]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 9 above)
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/expander.go:80]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 10 above)
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/expander.go:90]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/expander.go (hunk 8 above)
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/expander.go:70]: Thanks!

File: nifcloud/resources/computing/separateinstancerule/expander_test.go [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+		})
+	}
+}
+
+func TestExpandNiftyUpdateSeparateInstanceRuleInputForName(t *testing.T) {
+	r := New()
+	rd := schema.TestResourceDataRaw(t, r.Schema, map[string]interface{}{
+		"name": "test_separate_name",
+	})
+	rd.SetId("test_separate_name")
```
[the remark this answers is not in the document]
Comment (miyuush) [reply nifcloud/resources/computing/separateinstancerule/expander_test.go:89]: Yes, it is.
Thanks for the fix!

**Review by miyuush — APPROVED** [review]
Thanks for the fix!
LGTM!

**Pull request merged**